overlord/state/taskrunner_test.go

```diff
@@ -34,204 +35,291 @@ type taskRunnerSuite struct{}
 
 var _ = Suite(&taskRunnerSuite{})
 
-func (ts *taskRunnerSuite) TestEnsureTrivial(c *C) {
-	st := state.New(nil)
-
-	// setup the download handler
-	taskCompleted := sync.WaitGroup{}
-	r := state.NewTaskRunner(st)
-	fn := func(task *state.Task, tomb *tomb.Tomb) error {
-		task.State().Lock()
-		defer task.State().Unlock()
-		c.Check(task.Status(), Equals, state.DoStatus)
-		taskCompleted.Done()
-		return nil
-	}
-	r.AddHandler("download", fn)
-
-	// add a download task to the state tracker
-	st.Lock()
-	chg := st.NewChange("install", "...")
-	t := st.NewTask("download", "1...")
-	chg.AddTask(t)
-	taskCompleted.Add(1)
-	st.Unlock()
-
-	defer r.Stop()
-
-	// ensure just kicks the go routine off
-	r.Ensure()
-	taskCompleted.Wait()
-
-	st.Lock()
-	defer st.Unlock()
-	c.Check(t.Status(), Equals, state.DoneStatus)
-}
-
 type stateBackend struct {
-	runner *state.TaskRunner
+	mu           sync.Mutex
+	ensureBefore time.Duration
 }
 
-func (b *stateBackend) Checkpoint([]byte) error {
-	return nil
-}
+func (b *stateBackend) Checkpoint([]byte) error { return nil }
 
 func (b *stateBackend) EnsureBefore(d time.Duration) {
-	go b.runner.Ensure()
-}
-
-func (ts *taskRunnerSuite) TestEnsureComplex(c *C) {
-	b := &stateBackend{}
-	st := state.New(b)
-
-	r := state.NewTaskRunner(st)
-	b.runner = r
-
-	// setup handlers
-	orderingCh := make(chan string, 3)
-	fn := func(task *state.Task, tomb *tomb.Tomb) error {
-		task.State().Lock()
-		defer task.State().Unlock()
-		c.Check(task.Status(), Equals, state.DoStatus)
-		orderingCh <- task.Kind()
-		return nil
+	b.mu.Lock()
+	if d < b.ensureBefore {
+		b.ensureBefore = d
 	}
-	r.AddHandler("download", fn)
-	r.AddHandler("unpack", fn)
-	r.AddHandler("configure", fn)
-
-	defer r.Stop()
-
-	// run in a loop to ensure ordering is not correct by pure chance
-	for i := 0; i < 100; i++ {
-		st.Lock()
-		chg := st.NewChange("mock-install", "...")
-
-		// create sub-tasks
-		tDl := st.NewTask("download", "1...")
-		tUnp := st.NewTask("unpack", "2...")
-		tUnp.WaitFor(tDl)
-		chg.AddAll(state.NewTaskSet(tDl, tUnp))
-		tConf := st.NewTask("configure", "3...")
-		tConf.WaitFor(tUnp)
-		chg.AddAll(state.NewTaskSet(tConf))
-		st.Unlock()
+	b.mu.Unlock()
+}
 
-		// ensure just kicks the go routine off
-		// and then they get scheduled as they finish
+func ensureChange(c *C, r *state.TaskRunner, sb *stateBackend, chg *state.Change) {
+	sb.ensureBefore = 0
+	for sb.ensureBefore == 0 {
```

Comment:
something like this I think:

--- a/overlord/state/taskrunner_test.go
+++ b/overlord/state/taskrunner_test.go
@@ -53,6 +53,7 @@ func (b *stateBackend) EnsureBefore(d time.Duration) {
 func ensureChange(c *C, r *state.TaskRunner, sb *stateBackend, chg *state.Change) {
        sb.ensureBefore = 0
        for sb.ensureBefore == 0 {
+               sb.ensureBefore = time.Minute
                r.Ensure()
                r.Wait()
                chg.State().Lock()